### PR TITLE
Added table pagination support on front and backend

### LINKED ISSUE #197: Add Table Pagination support to backend for Users and Pitches
## Description

Add backend support for table pagination so that the frontend endpoints can paginate their data in the tables

## Learning objectives

- Objective

## Todo

These points are a rough guideline. **Please** feel free to discuss with others on the team about the best way to design this component!

- Todo

## Useful links

- Useful link

### PR DESCRIPTION
**Pull request by chenfeiyu132**
## Summary

Modified backend api endpoints to support pagination and filtering for fetching approved and pending as well as rejected users. Made frontend api calls also support queries that are passed in(which will be useful for filters and pages). Still need to complete the actual table pagination component
Closes #197 

## Changes

- Added processPagination and processFilters helper in the backend
- Added query string construction function in the front end to build uri
- Adapted filtering of data on the backend instead of front end for tables

## Requests / Responses

**Request**

GET `/api/users/all/pending?{query}` Returns a list of pending users given query(includes pagination)
GET `/api/users/all/approved?{query}` Returns a list of pending users given query(includes filtering and pagination)
GET `/api/users/all/rejected?{query}` Returns a list of denied users given query(includes filtering and pagination)

GET `/api/pitches/all/pending?{query}` Returns a list of pitches pending approval given query(includes filtering and pagination)
GET `/api/pitches/all/approved?{query}` Returns a list of approved pitches given query(includes filtering and pagination)

### CODE REVIEW

File: api/src/utils/user-utils.ts [hunk 1]
```diff
@@ -54,4 +55,56 @@ const getEditableFields = (currentUser: IUser, userId: string): UserKeys => {
...
+    const page = parseInt(req.query.page as string);
+    const limit = parseInt(req.query.limit as string);
+    const sortBy = req.query.sortBy as string;
+    const sortDirection = req.query.sortDirection as string;
+
+    if (sortBy) {
+      query.sort({ sortBy: sortDirection });
+    }
+
+    const skipIndex = (page - 1) * limit;
```
Comment (AndrewLester) [inline api/src/utils/user-utils.ts:99]: Do we want page to be indexed at 0 or 1? Not really a necessary change we should just consider both options.

File: client/src/api/builders.ts [hunk 2]
```diff
@@ -110,4 +110,30 @@ const buildLoginEndpoint = (endpoint: string): string =>
...
 const buildEndpoint = (...paths: string[]): string => paths.join('/');
 
-export { buildLoginEndpoint, get, post, put, del, isError, buildEndpoint };
+/**
+ * Builds the URI with complete params that are parsed correctly
+ * @param url the url with the endpoints built
+ * @param params the query params
+ * @returns the uri string with params appended correctly
+ */
+
```
Comment (AndrewLester) [inline client/src/api/builders.ts:119]:
```suggestion

```
Remove this empty line

File: client/src/pages/directory/index.tsx [hunk 3]
```diff
@@ -40,63 +44,46 @@ interface PaneWrapperProps {
...
   const [interests, setInterests] = useState<string[]>([]);
   const [teams, setTeams] = useState<string[]>([]);
   const [query, setQuery] = useState<string>('');
+  const [page, setPage] = useState<number>(1);
+  const [pageLimit, setPageLimit] = useState<number>(10);
+  const [params, setParams] = useState<Record<string, string[]>>({});
+
+  const getCurrUsers = useCallback(async (): Promise<void> => {
+    let res;
+    console.log(params);
```
Comment (AndrewLester) [inline client/src/pages/directory/index.tsx:57]: Remove console log

```suggestion

```

File: client/src/pages/directory/index.tsx [hunk 4]
```diff
@@ -40,63 +44,46 @@ interface PaneWrapperProps {
...
-      if (query.length === 0) {
-        return users;
-      }
-
-      const searchTerm = toLower(query.trim());
-      const queryParts = searchTerm.split(' ');
+    const newParams: Record<string, string[]> = {};
+    newParams['interests'] = interests;
+    newParams['teams'] = teams;
+    newParams['role'] = role === '' ? [] : [role];
```
Comment (AndrewLester) [inline client/src/pages/directory/index.tsx:81]: These object properties could be set inside the object literal.

**Review by AndrewLester — APPROVED** [review]
Everything's looking good, the front end paging table should be done shortly. Also, we may want to add a set of pagination state data to the responses that were changed here so that the client can confirm the page it's on and get the total number of items/pages.

**Review by amit-sawhney — APPROVED** [review]
I took some of all of this code and refactored it a bit and cleaned it up on my branch. I will integrate it with this frontend after it is merged.

File: api/src/api/pitch.ts [hunk 5]
```diff
@@ -18,6 +18,7 @@ import {
   approveClaim,
   declineClaim,
 } from '../utils/mailer-templates';
+import { processFilters, processPaignation } from '../utils/user-utils';
```
[the remark this answers is not in the document]
Comment (AndrewLester) [reply api/src/api/pitch.ts]: Seems to be fixed? I'm not sure how

**Pull request merged**